core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -60,29 +62,77 @@
   /**
    * Returns a new {@link ManifestReader} for an {@link InputFile}.
    * <p>
-   * <em>Note:</em> Most callers should use {@link #read(InputFile, Map)} to ensure that the
-   * schema used by filters is the latest table schema. This should be used only when reading a
-   * manifest without filters.
+   * <em>Note:</em> Most callers should use {@link #read(ManifestFile, FileIO, Map)} to ensure that
+   * manifest entries with partial metadata can inherit missing properties from the manifest metadata.
+   * <p>
+   * <em>Note:</em> Most callers should use {@link #read(InputFile, Map)} if all manifest entries
+   * contain full metadata and they want to ensure that the schema used by filters is the latest
+   * table schema. This should be used only when reading a manifest without filters.
    *
    * @param file an InputFile
    * @return a manifest reader
    */
   public static ManifestReader read(InputFile file) {
-    return new ManifestReader(file, null);
+    return new ManifestReader(file, null, InheritableMetadataFactory.empty());
   }
 
   /**
    * Returns a new {@link ManifestReader} for an {@link InputFile}.
+   * <p>
+   * <em>Note:</em> Most callers should use {@link #read(ManifestFile, FileIO, Map)} to ensure that
+   * manifest entries with partial metadata can inherit missing properties from the manifest metadata.
    *
    * @param file an InputFile
    * @param specsById a Map from spec ID to partition spec
    * @return a manifest reader
    */
   public static ManifestReader read(InputFile file, Map<Integer, PartitionSpec> specsById) {
-    return new ManifestReader(file, specsById);
+    return new ManifestReader(file, specsById, InheritableMetadataFactory.empty());
+  }
+
+  /**
+   * Returns a new {@link ManifestReader} for a {@link ManifestFile}.
+   *
+   * @param manifest a ManifestFile
+   * @param io a FileIO
+   * @param specsById a Map from spec ID to partition spec
+   * @return a manifest reader
+   */
+  public static ManifestReader read(ManifestFile manifest, FileIO io, Map<Integer, PartitionSpec> specsById) {
```

Comment:
Can we extract this API in a separated PR and merge first?

Reply:
I have implemented the feature flag locally. As soon as #738 is in, I’ll update this PR and I think we will merge it quickly as well. If this is already blocking your work, I can extract the API changes in ManifestReader. Let me know.

Reply:
It's in! Sorry for not getting it in sooner since it was blocking this one.

Reply:
@aokolnychyi , Thanks for the update. I have merged with some of this patch locally and it works well for now. Please take the time and go ahead with this PR. I may need some time to finish mine since we are on holiday season:)